include/ginkgo/core/multigrid/amgx_pgm.hpp

```diff
@@ -0,0 +1,175 @@
+class AmgxPgm : public EnableRestrictProlong<AmgxPgm<ValueType, IndexType>> {
+    friend class EnableRestrictProlong<AmgxPgm>;
+    friend class EnablePolymorphicObject<AmgxPgm, RestrictProlong>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Returns the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Returns the aggregate group.
+     *
+     * Aggregate group whose size is same as the number of rows stores the
+     * mapping information from row index to coarse row index.
+     * i.e., agg[row_idx] = coarse_row_idx.
+     *
+     * @return the aggregate group.
+     */
+    IndexType *get_agg() noexcept { return agg_.get_data(); }
+
+    /**
+     * @copydoc AmgxPgm::get_agg()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const IndexType *get_const_agg() const noexcept
+    {
+        return agg_.get_const_data();
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * The maximum number of iterations.
+         */
+        unsigned GKO_FACTORY_PARAMETER(max_iterations, 15u);
+
+        /**
+         * The maximum ratio of unassigned number, which is valid in the
+         * interval 0.0 ~ 1.0.
+         */
+        double GKO_FACTORY_PARAMETER(max_unassigned_percentage, 0.05);
+
+        /**
+         * Use the deterministic assign_to_exist_agg method or not.
+         *
+         * If the deterministic is true, always get the same aggregated group
+         * from the same matrix. Otherwise, the aggregated group might be
+         * different depending on the execution ordering.
+         */
+        bool GKO_FACTORY_PARAMETER(deterministic, false);
+    };
+    GKO_ENABLE_RESTRICT_PROLONG_FACTORY(AmgxPgm, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    void restrict_apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void prolong_applyadd_impl(const LinOp *b, LinOp *x) const override;
+
+    explicit AmgxPgm(std::shared_ptr<const Executor> exec)
+        : EnableRestrictProlong<AmgxPgm>(std::move(exec))
+    {}
+
+    explicit AmgxPgm(const Factory *factory,
+                     std::shared_ptr<const LinOp> system_matrix)
+        : EnableRestrictProlong<AmgxPgm>(factory->get_executor()),
+          parameters_{factory->get_parameters()},
+          system_matrix_{std::move(system_matrix)},
+          agg_(factory->get_executor(), system_matrix_->get_size()[0])
+    {
+        GKO_ASSERT(parameters_.max_unassigned_percentage <= 1.0);
+        GKO_ASSERT(parameters_.max_unassigned_percentage >= 0.0);
```

Comment:
Good question.

Reply:
it works because the percentage = 0 means to aggregate all pairs (under iter limit)
if it is equal to 1, the current implementation will run the match step once because only check the percent after the match step.

Reply:
okay, makes sense. thanks.